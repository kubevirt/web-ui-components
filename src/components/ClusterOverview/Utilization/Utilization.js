import React from 'react';
import { SparklineChart, Row, Col } from 'patternfly-react';
import PropTypes from 'prop-types';

import {
  DashboardCard,
  DashboardCardBody,
  DashboardCardHeader,
  DashboardCardTitle,
} from '../../Dashboard/DashboardCard';
import { ClusterOverviewContextGenericConsumer } from '../ClusterOverviewContext';

const UtilizationItem = ({ id, title, data, unit }) => (
  <div className="kubevirt-utilization__item">
    <Row>
      <Col lg={2} md={2} sm={2} xs={2}>
        <h2>{title}</h2>
      </Col>
      <Col lg={3} md={3} sm={3} xs={3}>
        <h2>{`${data[data.length - 1]} ${unit}`}</h2>
      </Col>
      <Col lg={7} md={7} sm={7} xs={7}>
        <SparklineChart
          id={id}
          data={{
            columns: [[unit, ...data]],
            unload: true,
            type: 'area',
          }}
          unloadBeforeLoad
          size={{ width: 430, height: 70 }}
        />
      </Col>
    </Row>
  </div>
);

UtilizationItem.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  data: PropTypes.array.isRequired,
  unit: PropTypes.string.isRequired,
};

class UtilizationBody extends React.PureComponent {
  render() {
    const { stats } = this.props;
    return Object.keys(stats).map(key => (
      <UtilizationItem id={key} key={key} title={stats[key].title} data={stats[key].data} unit={stats[key].unit} />
    ));
  }
}

UtilizationBody.propTypes = {
  stats: PropTypes.object.isRequired,
};

export const Utilization = ({ stats, loaded }) => (
  <DashboardCard>
    <DashboardCardHeader>
      <DashboardCardTitle>Cluster utilization</DashboardCardTitle>
    </DashboardCardHeader>
    <DashboardCardBody isLoading={!loaded}>
      <UtilizationBody stats={stats} />
    </DashboardCardBody>
  </DashboardCard>
);

Utilization.defaultProps = {
  loaded: false,
};

Utilization.propTypes = {
  stats: PropTypes.object.isRequired,
  loaded: PropTypes.bool,
};

const UtilizationConnected = () => (
  <ClusterOverviewContextGenericConsumer Component={Utilization} dataPath="utilizationStats" />
);

export default UtilizationConnected;
