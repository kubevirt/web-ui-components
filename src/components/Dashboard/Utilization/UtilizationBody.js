import React from 'react';
import PropTypes from 'prop-types';

export const UtilizationBody = ({ children }) => <div className="kubevirt-utilization__items">{children}</div>;

UtilizationBody.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
};
