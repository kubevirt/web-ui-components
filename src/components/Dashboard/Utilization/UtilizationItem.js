import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Measure from 'react-measure';

import { Col, Row } from 'patternfly-react';
import { ChartArea, ChartVoronoiContainer, Chart, ChartAxis, ChartTooltip } from '@patternfly/react-charts';

import { MEDIA_QUERY_SM } from '../../../utils';

import { InlineLoading } from '../../Loading';
import { NOT_AVAILABLE } from './strings';

export class UtilizationItem extends React.PureComponent {
  state = {
    dimensions: {
      width: -1,
    },
  };

  onResize = contentRect => {
    this.setState({ dimensions: contentRect.bounds });
  };

  render() {
    const { id, title, data, maxY, unit, isLoading, LoadingComponent } = this.props;
    const { width } = this.state.dimensions;

    const axis = {
      y: {
        show: true,
        padding: {
          top: 0,
          bottom: 0,
        },
        tick: {
          count: 3,
          format: v => `${v} ${unit}`,
        },
      },
      x: {
        show: false,
      },
    };
    if (maxY) {
      axis.y.max = maxY;
    }

    let actual;
    let chart = NOT_AVAILABLE;
    if (isLoading) {
      chart = <LoadingComponent />;
    } else if (data) {
      const chartData = data.map((val, index) => ({ x: index, y: val ? Number(val.toFixed(1)) : 0 }));
      actual = `${Math.round(data[data.length - 1])} ${unit}`;
      chart = (
        <Chart
          id={id}
          height={60}
          width={300}
          containerComponent={
            <ChartVoronoiContainer
              labels={datum => `${datum.y} ${unit}`}
              labelComponent={<ChartTooltip style={{ fontSize: 12, padding: 5 }} />}
            />
          }
          padding={{ top: 10, left: 40, bottom: 5 }}
          domainPadding={{ x: 0, y: 10 }}
        >
          <ChartArea data={chartData} />
          <ChartAxis dependentAxis tickValues={[0, maxY / 2, maxY]} style={{ tickLabels: { fontSize: 10 } }} />
        </Chart>
      );
    }

    let topClass;
    let rows;
    if (width < MEDIA_QUERY_SM) {
      topClass = 'kubevirt-utilization__item-narrow';
      rows = (
        <Fragment>
          <Row>
            <Col lg={6} md={6} sm={6} xs={6} className="kubevirt-utilization__item-narrow-title">
              {title}
            </Col>
            <Col className="kubevirt-utilization__item-actual item-actual--narrow" lg={6} md={6} sm={6} xs={6}>
              {actual}
            </Col>
          </Row>
          <Row>
            <Col className="kubevirt-utilization__item-chart kubevirt-utilization__item-chart--narrow">{chart}</Col>
          </Row>
        </Fragment>
      );
    } else {
      topClass = 'kubevirt-utilization__item';
      rows = (
        <Row>
          <Col lg={3} md={3} sm={3} xs={3}>
            {title}
          </Col>
          <Col className="kubevirt-utilization__item-actual" lg={2} md={2} sm={2} xs={2}>
            {actual}
          </Col>
          <Col
            className="kubevirt-utilization__item-chart kubevirt-utilization__item-chart--wide"
            lg={7}
            md={7}
            sm={7}
            xs={7}
          >
            {chart}
          </Col>
        </Row>
      );
    }

    return (
      <Measure bounds onResize={this.onResize}>
        {({ measureRef }) => (
          <div ref={measureRef} className={topClass}>
            {rows}
          </div>
        )}
      </Measure>
    );
  }
}

UtilizationItem.defaultProps = {
  maxY: null,
  data: null,
  LoadingComponent: InlineLoading,
  isLoading: false,
};

UtilizationItem.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  data: PropTypes.array,
  unit: PropTypes.string.isRequired,
  maxY: PropTypes.number,
  LoadingComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  isLoading: PropTypes.bool,
};
