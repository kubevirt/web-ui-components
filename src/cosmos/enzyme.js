import createContext from 'react-cosmos-test/enzyme';

import paths from '../../config/paths';

export default function createTestContext(args) {
  args.proxies = args.proxies || []; // workaround of react-cosmos-test bug
  return createContext({
    ...args,
    cosmosConfigPath: `${paths.config}/cosmos.config.js`,
  });
}
