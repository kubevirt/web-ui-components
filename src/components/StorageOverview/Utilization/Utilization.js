import React from 'react';
import PropTypes from 'prop-types';

import {
  DashboardCard,
  DashboardCardBody,
  DashboardCardHeader,
  DashboardCardTitle,
} from '../../Dashboard/DashboardCard';
import { StorageOverviewContext } from '../StorageOverviewContext';
import { InlineLoading } from '../../Loading';
import { UtilizationBody } from '../../Dashboard/Utilization/UtilizationBody';
import { UtilizationItem } from '../../Dashboard/Utilization/UtilizationItem';
import { getUtilizationVectorStats } from '../../../selectors';
import { formatBytes } from '../../../utils';

export const Utilization = ({
  iopsUtilization,
  latencyUtilization,
  throughputUtilization,
  recoveryRateUtilization,
  LoadingComponent,
}) => {
  const iopsStats = getUtilizationVectorStats(iopsUtilization);
  const latencyStats = getUtilizationVectorStats(latencyUtilization);

  const throughputStatsRaw = getUtilizationVectorStats(throughputUtilization);
  let throughputStats = null;
  let throughputMax = 0;
  let throughputMaxConverted;
  if (throughputStatsRaw) {
    throughputMax = Math.max(0, ...throughputStatsRaw);
    throughputMaxConverted = formatBytes(throughputMax);
    throughputStats = throughputStatsRaw.map(bytes => formatBytes(bytes, throughputMaxConverted.unit, 1).value);
  } else {
    throughputMaxConverted = formatBytes(throughputMax); // B
  }
  const throughputUnit = `${throughputMaxConverted.unit}/s`;

  const recoveryRateStatsRaw = getUtilizationVectorStats(recoveryRateUtilization);
  let recoveryRateStats = null;
  let recoveryRateMax = 0;
  let recoveryRateMaxConverted;
  if (recoveryRateStatsRaw) {
    recoveryRateMax = Math.max(0, ...recoveryRateStatsRaw);
    recoveryRateMaxConverted = formatBytes(recoveryRateMax);
    recoveryRateStats = recoveryRateStatsRaw.map(bytes => formatBytes(bytes, recoveryRateMaxConverted.unit, 1).value);
  } else {
    recoveryRateMaxConverted = formatBytes(recoveryRateMax); // B
  }
  const recoveryRateUnit = `${recoveryRateMaxConverted.unit}/s`;

  return (
    <DashboardCard>
      <DashboardCardHeader>
        <DashboardCardTitle>Performance</DashboardCardTitle>
      </DashboardCardHeader>
      <DashboardCardBody>
        <UtilizationBody>
          <UtilizationItem
            unit={throughputUnit}
            id="throughput"
            title="Throughput"
            data={throughputStats}
            LoadingComponent={LoadingComponent}
            isLoading={!throughputUtilization}
          />
          <UtilizationItem
            unit="IOPS"
            id="iops"
            title="IOPS"
            data={iopsStats}
            LoadingComponent={LoadingComponent}
            isLoading={!iopsUtilization}
          />
          <UtilizationItem
            unit="ms"
            id="latency"
            title="Latency"
            data={latencyStats}
            LoadingComponent={LoadingComponent}
            isLoading={!latencyUtilization}
          />
          <UtilizationItem
            unit={recoveryRateUnit}
            id="recoveryRate"
            title="Recovery rate"
            data={recoveryRateStats}
            LoadingComponent={LoadingComponent}
            isLoading={!recoveryRateUtilization}
          />
        </UtilizationBody>
      </DashboardCardBody>
    </DashboardCard>
  );
};

Utilization.defaultProps = {
  iopsUtilization: null,
  latencyUtilization: null,
  throughputUtilization: null,
  recoveryRateUtilization: null,
  LoadingComponent: InlineLoading,
};

Utilization.propTypes = {
  iopsUtilization: PropTypes.object,
  latencyUtilization: PropTypes.object,
  throughputUtilization: PropTypes.object,
  recoveryRateUtilization: PropTypes.object,
  LoadingComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
};

export const UtilizationConnected = () => (
  <StorageOverviewContext.Consumer>{props => <Utilization {...props} />}</StorageOverviewContext.Consumer>
);
