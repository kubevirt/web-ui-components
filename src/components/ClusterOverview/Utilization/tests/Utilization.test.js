import React from 'react';
import { render } from 'enzyme';

import { Utilization } from '../Utilization';
import { default as UtilizationFixtures } from '../fixtures/Utilization.fixture';

const testUtilizationOverview = () => <Utilization {...UtilizationFixtures.props} />;

describe('<Utilization />', () => {
  it('renders correctly', () => {
    const component = render(testUtilizationOverview());
    expect(component).toMatchSnapshot();
  });
});
