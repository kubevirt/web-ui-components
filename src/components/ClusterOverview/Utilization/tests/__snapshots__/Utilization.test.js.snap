// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Utilization /> renders correctly 1`] = `
<div
  class="card-pf kubevirt-dashboard__card"
>
  <div
    class="card-pf-heading kubevirt-dashboard__card-heading"
  >
    <h2
      class="card-pf-title"
    >
      Cluster utilization
    </h2>
  </div>
  <div
    class="card-pf-body"
  >
    <div
      class="kubevirt-utilization__item"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-2 col-md-2 col-sm-2 col-xs-2"
        >
          <h2>
            CPU
          </h2>
        </div>
        <div
          class="col-lg-3 col-md-3 col-sm-3 col-xs-3"
        >
          <h2>
            10 %
          </h2>
        </div>
        <div
          class="col-lg-7 col-md-7 col-sm-7 col-xs-7"
        >
          <div
            class=" chart-pf-sparkline"
          />
        </div>
      </div>
    </div>
    <div
      class="kubevirt-utilization__item"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-2 col-md-2 col-sm-2 col-xs-2"
        >
          <h2>
            I/O
          </h2>
        </div>
        <div
          class="col-lg-3 col-md-3 col-sm-3 col-xs-3"
        >
          <h2>
            20000 MBps
          </h2>
        </div>
        <div
          class="col-lg-7 col-md-7 col-sm-7 col-xs-7"
        >
          <div
            class=" chart-pf-sparkline"
          />
        </div>
      </div>
    </div>
    <div
      class="kubevirt-utilization__item"
    >
      <div
        class="row"
      >
        <div
          class="col-lg-2 col-md-2 col-sm-2 col-xs-2"
        >
          <h2>
            Latency
          </h2>
        </div>
        <div
          class="col-lg-3 col-md-3 col-sm-3 col-xs-3"
        >
          <h2>
            3.5 ms
          </h2>
        </div>
        <div
          class="col-lg-7 col-md-7 col-sm-7 col-xs-7"
        >
          <div
            class=" chart-pf-sparkline"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
